Updated Intune integration to work with Microsoft Graph

trs80 opened the pull request:
# Description
Updates the Intune cloud integration to work with MS Graph API as well as Azure AD API

# Impacts
Makes Intune SCEP support work for new installs since Azure AD API is deprecated and going away soonish

# Issue
fixes #6933

# Delete branch after merge
NO

# Checklist
(REQUIRED) - [yes, no or n/a]
- [yes] Document the feature
- [no] Add unit tests
- [no] Add acceptance tests (TestLink)

# NEWS file entries
## Bug Fixes
If an issue exists on Github, please refer to it (name) along with it's number...
* Updates Intune integration to new Microsoft Graph API (#6933)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/inverse-inc/packetfence?pullRequest=7115) <br/>All committers have signed the CLA.

fdurand commented on go/caddy/pfpki/cloud/intune.go:
```diff
@@ -161,10 +164,31 @@ func (cl *Intune) NewCloud(ctx context.Context, name string) error {
 	cl.Client = client
 
-	req, err := http.NewRequest("GET", graphRequest, nil)
+	var aadGraphFailed bool
```
it should be msGraphFailed instead since the first request is for msGraphResourceUrl

fdurand commented on go/caddy/pfpki/cloud/intune.go:
```diff
@@ -161,10 +164,31 @@ func (cl *Intune) NewCloud(ctx context.Context, name string) error {
 	if err != nil {
-		log.Print(err)
-		os.Exit(1)
+		aadGraphFailed = true
```
msGraphFailed

fdurand commented on go/caddy/pfpki/cloud/intune.go:
```diff
@@ -161,10 +164,31 @@ func (cl *Intune) NewCloud(ctx context.Context, name string) error {
+		aadGraphFailed = true
+	}
+
+	if aadGraphFailed {
```
msGraphFailed

fdurand commented on go/caddy/pfpki/cloud/intune.go:
```diff
@@ -195,15 +219,8 @@ func (cl *Intune) NewCloud(ctx context.Context, name string) error {
 			for _, n := range v.([]interface{}) {
 				for a, b := range n.(map[string]interface{}) {
-					if a == "serviceName" {
+					if a == "providerName" && !aadGraphFailed || a == "serviceName" && aadGraphFailed {
```
msGraphFailed

fdurand commented:
For the screenshot can we add both or at least say if Azure AD API is deprecated then do that ?

trs80 commented:
For the screenshot, Azure AD Graph is so deprecated you can't even add it via the web UI any more, and will stop working after the end of 2022. So I've added a note saying that and to add the new permissions instead.

I've changed the variable name, the Azure AD Graph codepath does need testing by someone else who already has it set up.

trs80 commented:
Fixed conflict with upstream (adding spew.Dump(err.Error()) basically)

fdurand commented:
@trs80 i did another branch for that since adal will be deprecated in favor of azidentity https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/azidentity/MIGRATION.md#autorestadal-5 .
I also completly deprecated Azure AD API and if you want to use intune with new version of PacketFence then you will need to adjust the permissions.

https://github.com/inverse-inc/packetfence/compare/fix/intune

Btw i don´t want to rob your work so if you want you can take the code i did and apply it on your PR.

trs80 commented:
I'm not fussed about credit, I just want a working version merged. I can update the docs on top of your PR if you want.

fdurand commented:
Other branch has been merged in devel.
Thanks a lot for your work.
